src/ppx/xr_component.cpp

```diff
@@ -0,0 +1,395 @@
+            case XR_SESSION_LOSS_PENDING:
+            case XR_FRAME_DISCARDED:
+                mShouldRender = false;
+                break;
+
+            default:
+                PPX_ASSERT_MSG(false, "xrBeginFrame failed!");
+                break;
+        }
+    }
+}
+
+void XrComponent::EndFrame()
+{
+    mCompositionLayerProjection.space     = mRefSpace;
+    mCompositionLayerProjection.viewCount = static_cast<uint32_t>(GetViewCount());
+    mCompositionLayerProjection.views     = mCompositionLayerProjectionViews.data();
+    XrCompositionLayerBaseHeader* layer   = mShouldRender ? reinterpret_cast<XrCompositionLayerBaseHeader*>(&mCompositionLayerProjection) : nullptr;
+
+    XrFrameEndInfo frameEndInfo = {
+        .type                 = XR_TYPE_FRAME_END_INFO,
+        .displayTime          = mFrameState.predictedDisplayTime,
+        .environmentBlendMode = mBlend,
+        .layerCount           = static_cast<uint32_t>(mShouldRender ? 1 : 0),
+        .layers               = &layer,
+    };
+
+    VERIFY_XR_CALL(xrEndFrame(mSession, &frameEndInfo));
+}
+
+glm::mat4 XrComponent::GetCurrentViewViewMatrix() const
+{
+    PPX_ASSERT_MSG((mCurrentViewIndex < mViews.size()), "Invalid view index!");
+    const XrView&  view = mViews[mCurrentViewIndex];
+    const XrPosef& pose = view.pose;
+    // OpenXR is using right handed system which is the same as Vulkan
+    glm::quat quat         = glm::quat(pose.orientation.w, pose.orientation.x, pose.orientation.y, pose.orientation.z);
+    glm::mat4 rotation     = glm::mat4_cast(quat);
+    glm::vec3 position     = glm::vec3(pose.position.x, pose.position.y, pose.position.z);
+    glm::mat4 translation  = glm::translate(glm::mat4(1.f), position);
+    glm::mat4 view_glm     = translation * rotation;
+    glm::mat4 view_glm_inv = glm::inverse(view_glm);
+    return view_glm_inv;
+}
+
+glm::mat4 XrComponent::GetCurrentViewProjectionMatrix(float nearZ, float farZ) const
+{
+    PPX_ASSERT_MSG((mCurrentViewIndex < mViews.size()), "Invalid view index!");
+    const XrView& view = mViews[mCurrentViewIndex];
+    const XrFovf& fov  = view.fov;
+
+    const float tan_left  = tanf(fov.angleLeft);
+    const float tan_right = tanf(fov.angleRight);
+
+    const float tan_down = tanf(fov.angleDown);
+    const float tan_up   = tanf(fov.angleUp);
+
+    const float tan_width  = tan_right - tan_left;
+    const float tan_height = (tan_up - tan_down);
+
+    const float a00 = 2 / tan_width;
+    const float a11 = 2 / tan_height;
+
+    const float a20 = (tan_right + tan_left) / tan_width;
+    const float a21 = (tan_up + tan_down) / tan_height;
+    const float a22 = -farZ / (farZ - nearZ);
+
+    const float a32 = -(farZ * nearZ) / (farZ - nearZ);
+
+    const float mat[16] = {
+        a00,
+        0,
+        0,
+        0,
+        0,
+        a11,
+        0,
+        0,
+        a20,
+        a21,
+        a22,
+        -1,
+        0,
+        0,
+        a32,
+        0,
+    };
```

Comment:
Can you format the matrix in a 4x4 fashion? You can use 

``
// clang-format off
const float mat[16] = {
   ...
}
// clang-format on
```

Reply:
fixed